### PR TITLE
[RFR] Replace DependsOn Component

### LINKED ISSUE #1742: DependsOn not working
**What you were expecting:**
`<DependsOn>` is working as described in documentation.

**What happened instead:**
`<DependsOn>` is causing error and blank page.

**Related code:**
```
<DependsOn source="name" value="abstract">
  <TextField source="name">
</DependsOn>
```
applied in *next* branch demo project.

**Environment**

* react-admin: 2.0.0-beta4
* React: 16.2.0
dependson_error

### PR DESCRIPTION
**Pull request by fzaninotto**
`<DependsOn>` currently works only for Create and Edit views. It can be replaced by a more lightweight solution using render props:

```diff
- import { DependentInput } from 'aor-dependent-input';
+ import { FormDataConsumer } from 'react-admin';

export const UserCreate = (props) => (
    <Create {...props}>
        <SimpleForm>
            <TextInput source="firstName" />
            <TextInput source="lastName" />
            <BooleanInput source="hasEmail" label="Has email ?" />
-           <DependentInput dependsOn="hasEmail">
-                <TextInput source="email" />
-           </DependentInput>
+           <FormDataConsumer>
+               {(formData, ...rest) => formData.hasEmail && 
+                   <TextInput source="email" {...rest} />
+               }
+           </FormDataConsumer>
        </SimpleForm>
    </Create>
);
```

As for the Show view, the user should grab the record by using the `<ShowController>/<ShowView>` syntax instead of `<Show>`:

```diff
export const PostShow = props => (
-    <Show {...props}>
+    <ShowController {...props}>
+        {controllerProps => (
+            <ShowView {...props} {...controllerProps}>
                <SimpleShowView>
                        <TextField source="title" />
+                      {controllerProps.record && controllerProps.record.title &&
                                <TextField source="teaser" />
+                       }
                </SimpleShowView>
+            </ShowView>
+        )}
+    </ShowController>
-    </Show>
);
```

Closes #1742 

- [x] Remove `DependsOn` component, tests, and doc
- [x] Add `FormDataConsumer` component and doc
- [x] Update upgrade guide for conditional inputs
- [x] Document conditional fields in `Show` view
- [x] Update upgrade guide for conditional fields

### CODE REVIEW

File: UPGRADE.md [hunk 1]
```diff
@@ -994,14 +997,16 @@ export const UserCreate = (props) => (
             <TextInput source="lastName" />
             <BooleanInput source="hasEmail" label="Has email ?" />
 -           <DependentInput dependsOn="hasEmail">
-+           <DependsOn source="hasEmail">
```
Comment (fzaninotto) [inline UPGRADE.md]: don't trust GitHub's diff, there is no trace of `<DependsOn>` in the last version.

**Comment by fzaninotto** [comment]
Switching to RFR

File: UPGRADE.md [hunk 2]
```diff
@@ -979,13 +980,17 @@ Update your `import` statements accordingly:
...
 
-## aor-dependent-input integrated into core
+## `aor-dependent-input` Was Removed
 
-The `DependentInput` and `DependentField` components of `aor-dependent-input` have been merged into one named `DependsOn`. Besides, the `dependsOn` prop was renamed to `source`. Finally, this component is now bundled by default in the main `react-admin` package. 
+The `aor-dependent-input` package has been removed.
+
+You can achieve a similar effect to the old `<DependentInput>` component by using the new `<FormDataConsumer>` component.
+
+To display a component based on the value of the current (edited) record, wrap that component with `<FormDataConsumer>`, which uses grabs the form data from the redux-form state, and passes it to a child function:
```
Comment (jpetitcolas) [inline UPGRADE.md]: s/uses grabs/grabs/

File: packages/ra-core/src/form/FormDataConsumer.js [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ * ); 
+ * 
+ * @example
+ * 
+ * const OrderEdit = (props) => (
+ *     <Edit {...props}>
+ *         <SimpleForm>
+ *             <SelectInput source="country" choices={countries} />
+ *             <FormDataConsumer>
+ *                 {(formData, ...rest) =>
```
Comment (jpetitcolas) [inline packages/ra-core/src/form/FormDataConsumer.js]: :+1:

**Pull request merged**